gluon-config-mode-contact-info: an "obligatory" option in site.conf

rubo77 opened the pull request:
We already have a useable stand-alone packet to replace the `gluon-config-mode-contact-info` package in our repository at:

https://github.com/freifunk-kiel/ffki-packages/tree/master/gluon-config-mode-contact-obligatory

kb-light commented:
First `package/gluon-config-mode-contact-info/files/lib/gluon/config-mode/wizard/0500-contact-info.lua` can actually be found as: `package/gluon-config-mode-contact-info/luasrc/lib/gluon/config-mode/wizard/0500-contact-info.lua`.

Second, I think the fact that the owner information is public is a reason why you do not want to have this as "obligatory" but let the people decide on their own.

rubo77 commented:
obligatory in this package only means, that you have to fill in **anything** (even just a space is ok if you want to stay anonymous)

rotanid commented on package/gluon-config-mode-contact-info/check_site.lua:
```diff
@@ -0,0 +1 @@
+need_boolean('owner.obligatory', false)
```
every text file should end with a newline character, maybe you can add one here (github also highlights the missing newline with a red icon)

rotanid commented:
nice @rubo77 , we're planning on doing something similar, having a "are you sure you want to leave the contact field empty? ..." message appear, so the people installing nodes at least don't forget it by accident.
forcing will probably result in many contact-informations that are wrong on purpose...
what do you think?

@kb-light well, one could argue the PPA specifices you have to be reachable as a node operator..

rotanid commented:
@rubo77 regarding what @kb-light said: you put the 0500-contact-info.lua in the wrong folder, it switched places a while ago, please edit it in the new location rather than adding it to the old location.

rubo77 commented:
@rotanid, good idea. Did you create a packet for this already? I'll add the standalone packet, we already tested in the first message later.

We assume, that at the moment many users keep the contact field empty, cause of lazyness or they want to fill it out later, but never find the time. only a small fraction doesn't want to be reacheable.

I also thought of an option like:

[x] I would like to be notified, if my node is offline
> ... tell us how to reach you by the contact field or a private email or an online nodewather system, that will be available soon

I think, if a popup asks for confirmation to leave the field empty is not strong enough. If the field is obligatory, we get a better hint in the `json` string, that the user really wants to be anonymous, and didn't just forget or postponed to enter anything.

rotanid commented on package/gluon-config-mode-contact-info/files/lib/gluon/config-mode/wizard/0500-contact-info.lua:
```diff
@@ -0,0 +1,35 @@
+local cbi = require "luci.cbi"
```
this whole file is in the wrong folder, a while ago it switched places, maybe you started your development based on an older revision of gluon.

jplitza commented:
@rubo77 Such a checkbox doesn't really make sense, because to spam an address you have to have confirmed opt-in by that person, meaning that you also verified the opt-in really came from the owner of the address. So if you have a monitoring solution, simply let it send an "You're node went down for the first time, do you want to receive such notices in the future?". (I think you're using my nodewatcher? That still has to be implemented there.)

rubo77 commented:
I moved `files/lib/gluon/config-mode/wizard/0500-contact-info.lua` to `luasrc/lib/gluon/config-mode/wizard/0500-contact-info.lua` and added the ending linebreak.

anything else?

kb-light commented on package/gluon-config-mode-contact-info/luasrc/lib/gluon/config-mode/wizard/0500-contact-info.lua:
```diff
@@ -1,21 +1,22 @@
   local o = s:option(cbi.Value, "_contact", i18n.translate("Contact info"))
   o.default = uci:get_first("gluon-node-info", "owner", "contact", "")
-  o.rmempty = true
+  o.rmempty = not site.owner.obligatory
```
Have you tested this in case `owner` is not set within site.conf

rubo77 replied:
yes

kb-light replied:
if `obligatory` is not set everything works fine, but in case `owner` is not set, there will be an error: ` attempt to index field 'owner' (a nil value)`
So you should change this to: `o.rmempty = not (site.owner or {}).obligatory`

rubo77 replied:
ok, since I refactored the option to `site.config_mode.owner.obligatory`  I changed this to 

      o.rmempty = not ((site.config_mode or {}).owner or {}).obligatory

kb-light commented on package/gluon-config-mode-contact-info/luasrc/lib/gluon/config-mode/wizard/0500-contact-info.lua:
```diff
@@ -1,21 +1,22 @@
   local s = form:section(cbi.SimpleSection, nil, i18n.translate(
-    'You can provide your contact information here to '
-      .. 'allow others to contact you. Please note that '
+    'Please provide your contact information here to '
```
In my opinion, you should not change the default but may add an alternative text which is shown in case `obligatory==true`.

rubo77 replied:
doesn't that make the code unnecesary complicated?

the new text fits both options.

rotanid replied:
i agree that the changed text fits both cases

kb-light commented on docs/site-example/site.conf:
```diff
@@ -190,4 +190,9 @@
   --   },
   -- },
+
+  -- define if the contact field is obligatory
```
this should be:
`-- define if the contact field is obligatory (optional)

rubo77 replied:
why? all other optional settings don't have the hint (optional) I think the comments in front make this obvious already

kb-light replied:
It seems to be not consistent how to highlight optional settings, so maybe this is fine as it is.

kb-light commented on package/gluon-config-mode-contact-info/README.md:
```diff
@@ -0,0 +1,8 @@
+gluon-config-mode-contact-info
```
Please move the documentation to `docs/package/` and try to use a similar style like the existing ones.

rotanid commented on docs/site-example/site.conf:
```diff
@@ -181,13 +181,18 @@
   -- Skip setup mode (config mode) on first boot
   -- setup_mode = {
-  --  skip = true,
+  --   skip = true,
```
there really is a space missing, but that has nothing to do with this pull request

rubo77 replied:
should I revert this correction then? or just leave it, as minor, uninteresting side-fix ;-)

kb-light commented on docs/site-example/site.conf:
```diff
@@ -181,13 +181,18 @@
+    -- owner = {
+      -- obligatory = true
+    -- },
+  --},
```
You missed a space here

rubo77 replied:
corrected!

NeoRaider commented on package/gluon-config-mode-contact-info/Makefile:
```diff
@@ -13,10 +13,15 @@ PKG_CONFIG_DEPENDS += $(GLUON_I18N_CONFIG)
   SECTION:=gluon
   CATEGORY:=Gluon
-  TITLE:=Set a custom string that will be distributed in the mesh.
+  TITLE:=Set a custom string for a contact info that will be distributed in the mesh.
```
"for a contact info" sounds weird, "information" is always plural. Also the whole sentence doesn't make a lot of sense.

Maybe "Allows the user to provide contact information to be distributed in the mesh"

NeoRaider commented on package/gluon-config-mode-contact-info/Makefile:
```diff
@@ -13,10 +13,15 @@ PKG_CONFIG_DEPENDS += $(GLUON_I18N_CONFIG)
 
+define Package/gluon-config-mode-contact-info/description
+	Set a custom string for a contact info that will be distributed in the mesh.
+	Optionally definable obligatory in site.conf
```
"Can be made obligatory in site.conf"

NeoRaider commented on package/gluon-config-mode-contact-info/check_site.lua:
```diff
@@ -0,0 +1,3 @@
+if need_table('config_mode', nil, false) and need_table('config_mode.owner', nil, false) then
+  need_boolean('config_mode.geo_location.obligatory', false)
```
The above string has a copy&paste mistake.

NeoRaider commented on package/gluon-config-mode-contact-info/i18n/de.po:
```diff
@@ -14,12 +14,12 @@ msgid "Contact info"
 msgstr ""
-"Hier kannst du einen <em>öffentlichen</em> Hinweis hinterlegen, um anderen "
-"zu ermöglichen, Kontakt mit dir aufzunehmen. Bitte beachte, dass "
+"Hier bitte einen <em>öffentlichen</em> Hinweis hinterlegen, um anderen "
```
Dies ist kein ganzer Satz, bitte korrigieren.

rubo77 commented:
Ich hab den Satz noch mal überarbeitet: 

    +"Bitte hinterlege hier einen <em>öffentlichen</em> Hinweis, um anderen "
    +"zu ermöglichen, Kontakt mit dir aufzunehmen. Beachte, dass "
    +"dieser Hinweis zusammen mit den Koordinaten deines Knotens öffentlich "
    +"im Internet einsehbar sein wird."

NeoRaider commented:
Das klingt jetzt seltsam, als könnte man auch einen nicht öffentlichen Hinweis hinterlegen... Vielleicht das erste "öffentlich" streichen und statt dessen das zweite (oder sogar "öffentlich im Internet") hervorheben?

rubo77 commented:
```
"Bitte hinterlege hier einen Hinweis, wie andere mit dir Kontakt aufnehmen "
"können. Beachte dass dieser Hinweis zusammen mit den Koordinaten deines "
"Knotens <em>öffentlich</em> im Internet einsehbar sein wird."
```

NeoRaider commented on package/gluon-config-mode-contact-info/Makefile:
```diff
@@ -13,10 +13,15 @@ PKG_CONFIG_DEPENDS += $(GLUON_I18N_CONFIG)
 endef
 
+define Package/gluon-config-mode-contact-info/description
+	Set a custom string for a contact info that will be distributed in the mesh.
```
You updated the TITLE, but left the "a contact info" in the description.

NeoRaider commented on package/gluon-config-mode-contact-info/i18n/de.po:
```diff
@@ -14,14 +14,13 @@ msgid "Contact info"
-"zu ermöglichen, Kontakt mit dir aufzunehmen. Bitte beachte, dass "
-"dieser Hinweis auch öffentlich im Internet, zusammen mit den Koordinaten "
-"deines Knotens, einsehbar sein wird."
+"Bitte hinterlege hier einen Hinweis, wie andere mit dir Kontakt aufnehmen "
```
Der Satz gefällt mir noch immer nicht, das "wie" am Anfang des Nebensatzes klingt nicht sehr schön. Wollen wir den Nebensatz einfach mit "um ..." lassen, wir er war?

rubo77 commented:
OK, so?

"Bitte hinterlege hier einen Hinweis, um anderen "
"zu ermöglichen, Kontakt mit dir aufzunehmen. Beachte, dass "
"dieser Hinweis auch <em>öffentlich</em> im Internet, zusammen mit den Koordinaten "
"deines Knotens, einsehbar sein wird."

rubo77 commented:
Ich würde hier auch gerne Chat-Nickname ergänzen:

    "e.g. E-mail, chat-nickname or phone number"

Aber dies besser in einem extra PR oder?
